packages/flutter/lib/src/painting/text_span.dart

```diff
@@ -196,78 +202,113 @@ class TextSpan extends DiagnosticableTree {
 
   /// Walks this text span and its descendants in pre-order and calls [visitor]
   /// for each span that has text.
-  bool visitTextSpan(bool visitor(TextSpan span)) {
+  @override
+  bool visitChildren(InlineSpanVisitor visitor) {
     if (text != null) {
       if (!visitor(this))
         return false;
     }
     if (children != null) {
-      for (TextSpan child in children) {
-        if (!child.visitTextSpan(visitor))
+      for (InlineSpan child in children) {
+        if (!child.visitChildren(visitor))
           return false;
       }
     }
     return true;
   }
 
   /// Returns the text span that contains the given position in the text.
-  TextSpan getSpanForPosition(TextPosition position) {
+  @override
+  InlineSpan getSpanForPosition(TextPosition position) {
     assert(debugAssertIsValid());
-    final TextAffinity affinity = position.affinity;
-    final int targetOffset = position.offset;
-    int offset = 0;
+    final TrackingInt offset = TrackingInt();
     TextSpan result;
-    visitTextSpan((TextSpan span) {
-      assert(result == null);
-      final int endOffset = offset + span.text.length;
-      if (targetOffset == offset && affinity == TextAffinity.downstream ||
-          targetOffset > offset && targetOffset < endOffset ||
-          targetOffset == endOffset && affinity == TextAffinity.upstream) {
-        result = span;
+    visitChildren((InlineSpan span) {
+      result = span.getSpanForPositionVisitor(position, offset);
+      if (result != null) {
         return false;
       }
-      offset = endOffset;
       return true;
     });
     return result;
   }
+  @override
+  InlineSpan getSpanForPositionVisitor(TextPosition position, TrackingInt offset) {
+    if (text == null) {
+      return null;
+    }
+    final TextAffinity affinity = position.affinity;
+    final int targetOffset = position.offset;
+    final int endOffset = offset.value + text.length;
+    if (offset.value == targetOffset && affinity == TextAffinity.downstream ||
+        offset.value < targetOffset && targetOffset < endOffset ||
+        endOffset == targetOffset && affinity == TextAffinity.upstream) {
+      return this;
+    }
+    offset.value = endOffset;
+    return null;
+  }
 
   /// Flattens the [TextSpan] tree into a single string.
   ///
   /// Styles are not honored in this process. If `includeSemanticsLabels` is
   /// true, then the text returned will include the [semanticsLabel]s instead of
   /// the text contents when they are present.
-  String toPlainText({bool includeSemanticsLabels = true}) {
-    assert(debugAssertIsValid());
+  ///
+  /// When [includePlaceholders] is true, [PlaceholderSpan]s in the tree will be
+  /// represented as a 0xFFFC 'object replacement character'.
+  @override
+  String toPlainText({bool includeSemanticsLabels = true, bool includePlaceholders = true}) {
```

Comment:
looks like your implementations of toPlainText, codeUnitAt, etc, should be in the superclass? (with the implementations of compute* in the subclasses?)